apps/openmw/mwworld/weather.cpp

```diff
@@ -101,6 +104,95 @@ float Weather::cloudBlendFactor(float transitionRatio) const
     return transitionRatio / mCloudsMaximumPercent;
 }
 
+RegionWeather::RegionWeather(const ESM::Region& region)
+    : mWeather(invalidWeatherID)
+    , mChances()
+{
+    mChances.reserve(10);
+    mChances.push_back(region.mData.mClear);
+    mChances.push_back(region.mData.mCloudy);
+    mChances.push_back(region.mData.mFoggy);
+    mChances.push_back(region.mData.mOvercast);
+    mChances.push_back(region.mData.mRain);
+    mChances.push_back(region.mData.mThunder);
+    mChances.push_back(region.mData.mAsh);
+    mChances.push_back(region.mData.mBlight);
+    mChances.push_back(region.mData.mA);
+    mChances.push_back(region.mData.mB);
+}
+
+RegionWeather::RegionWeather(const ESM::RegionWeatherState& state)
+    : mWeather(state.mWeather)
+    , mChances(state.mChances)
+{
+}
+
+RegionWeather::operator ESM::RegionWeatherState() const
+{
+    ESM::RegionWeatherState state =
+        {
+            mWeather,
+            mChances
+        };
+
+    return state;
+}
+
+void RegionWeather::setChances(const std::vector<char>& chances)
+{
+    if(mChances.size() < chances.size())
+    {
+        mChances.reserve(chances.size());
+    }
+
+    std::vector<char>::const_iterator it = chances.begin();
+    for(size_t i = 0; it != chances.end(); ++it, ++i)
+    {
+        mChances[i] = *it;
+    }
+
+    // Regional weather no longer supports the current type, select a new weather pattern.
+    if((static_cast<size_t>(mWeather) >= mChances.size()) || (mChances[mWeather] == 0))
+    {
+        chooseNewWeather();
+    }
+}
+
+void RegionWeather::setWeather(int weatherID)
+{
+    mWeather = weatherID;
+}
+
+int RegionWeather::getWeather()
+{
+    // If the region weather was already set (by ChangeWeather, or by a previous call) then just return that value.
+    // Note that the region weather will be expired periodically when the weather update timer expires.
+    if(mWeather == invalidWeatherID)
+    {
+        chooseNewWeather();
+    }
+
+    return mWeather;
+}
+
+void RegionWeather::chooseNewWeather()
+{
+    // All probabilities must add to 100 (responsibility of the user).
+    // If chances A and B has values 30 and 70 then by generating 100 numbers 1..100, 30% will be lesser or equal 30
+    // and 70% will be greater than 30 (in theory).
+    int chance = Misc::Rng::rollDice(100) + 1; // 1..100
```

Comment:
I didn't modify that code much, it was lifted from the original algorithm as is.

Technically, the user is required to provide values such that the sum is 100, otherwise it has erratic behavior according to various sources (MWSFD, UESP). I didn't investigate the original's erratic behavior myself, so I can't say if your suggestion is correct.